src/wayback/_http.py

```diff
@@ -0,0 +1,591 @@
+                del headers['content-encoding']
+                headers['Content-Encoding'] = 'gzip'
+            return super().from_httplib(httplib_response, **response_kwargs)
+
+    HTTPConnectionPool.ResponseCls = WaybackUrllib3Response
+else:
+    # urllib3 v2.x:
+    #
+    # Unfortunately, we can't wrap the `HTTPConnection.getresponse` method in
+    # urllib3 v2, since it may have read the response body before returning. So
+    # we patch the HTTPHeaderDict class instead.
+    from urllib3._collections import HTTPHeaderDict as Urllib3HTTPHeaderDict
+    _urllib3_header_init = Urllib3HTTPHeaderDict.__init__
+
+    def _new_header_init(self, headers=None, **kwargs):
+        if headers:
+            if isinstance(headers, (Urllib3HTTPHeaderDict, dict)):
+                pairs = list(headers.items())
+            else:
+                pairs = list(headers)
+            if (
+                ('content-encoding', '') in pairs and
+                ('Content-Encoding', 'gzip') in pairs
+            ):
+                headers = [pair for pair in pairs
+                           if pair[0].lower() != 'content-encoding']
+                headers.append(('Content-Encoding', 'gzip'))
+
+        return _urllib3_header_init(self, headers, **kwargs)
+
+    Urllib3HTTPHeaderDict.__init__ = _new_header_init
+# END HACK
+#####################################################################
+
+
+class WaybackHttpResponse:
+    """
+    Represents an HTTP response from a server. This might be included as an
+    attribute of an exception, but should otherwise not be exposed to user
+    code in normal circumstances. It's meant to wrap to provide a standard,
+    thread-safe interface to response objects from whatever underlying HTTP
+    tool is being used (e.g. requests, httpx, etc.).
+    """
+    status_code: int
+    headers: dict
+    encoding: Optional[str] = None
+    url: str
+    links: dict
+
+    def __init__(self, url: str, status_code: int, headers: dict, links: dict = None, encoding: str = None):
+        self.url = url
+        self.status_code = status_code
+        self.headers = headers
+        self.links = links or {}
+        self.encoding = encoding
+
+    @property
+    def redirect_url(self) -> str:
+        """
+        The absolute URL this response redirects to. It will always be a
+        complete URL with a scheme and host. If the response is not a redirect,
+        this returns an empty string.
+        """
+        if self.status_code >= 300 and self.status_code < 400:
+            location = self.headers.get('location')
+            if location:
+                return urljoin(self.url, location)
+
+        return ''
+
+    @property
+    def is_success(self) -> bool:
+        """Whether the status code indicated success (2xx) or an error."""
+        return self.status_code >= 200 and self.status_code < 300
+
+    @property
+    def is_memento(self) -> bool:
+        return 'Memento-Datetime' in self.headers
```

Comment:
This property is not as generic as everything else in this class (which sticks to “HTTP” as the level of abstraction), but it was very helpfully convenient to add here. If we wanted to be more strict we could pull it back out into a utility function.